WordPress/gutenberg
# Layout: Move layout definitions out of theme.json

## What?


This is an experimental PR digging into an idea from https://github.com/WordPress/gutenberg/pull/50268#discussion_r1184720689 to remove the layout definitions from `theme.json`.

The approach in this PR is to try to generally retain the same shape and behaviour of the layout definitions, but switch where we retrieve the definitions _from_. They are now sourced from a PHP function where they're hard-coded, and in JS, they're stored in a `const` that is imported where appropriate.

This PR also looks to try to resolve https://github.com/WordPress/gutenberg/issues/50011 by calling `getLayoutStyles` from the post editor.

## Why?


As raised in https://github.com/WordPress/gutenberg/pull/50268#discussion_r1184720689, the layout definitions belong to core and are not intended to be changed or overridden by consumers. It is still useful to have central places where those definitions are stored. This PR initially proposed storing the layout definitions in a single PHP location, but has since evolved to include a corresponding JS constant.



## How?


* Move layout definitions from `theme.json` to a `gutenberg_get_layout_definitions` function, and to a `const` in JS.
* Call `gutenberg_get_layout_definitions` wherever we need the definitions:
  * In the Theme JSON class when outputting layout rules
  * In `layout.php` when outputting classnames and rules at the individual block level
* In the block editor code, add a JS const for the layout definitions, and import it where appropriate / use it as a default value, and remove lookups from `theme.json` settings
* Export `getLayoutStyles` as a private export
* Import it in the post editor, and use it to output base layout styles when the user has opted out of displaying theme styles in the editor

## Testing Instructions





This PR refactors both PHP and JS output for layout, so it'd be good to test all of the following to make sure nothing was missed:

* Each Group block variation (Group, Row, Stack, Grid) in the post editor, site editor, and site frontend. Test that default block spacing is output correctly, and that block spacing output is correct when set at the individual block level.
* Adjust global block spacing in the site site editor, adjust block-level spacing in global styles (e.g. block spacing for all social icons blocks), ensure output is consistent in site editor and on site frontend.
* Check that default layout styles are still output correctly in the post editor and site frontend for Classic themes (e.g. the social icons and columns blocks still visually have spacing)

## Screenshots or screencast
